naming/pubsub.md

```diff
@@ -0,0 +1,95 @@
+When a node attempts to retrieve an IPNS record from the network it uses the router to query for the IPNS record(s) associated with the IPNS key; the node then validates the received records.
+
+In this spec we address building a router based on a PubSub system, particularly focusing on libp2p PubSub.
+
+# PubSub Protocol Overview
+
+The protocol has four components:
+- [IPNS Records and Validation](/README.md)
+- [libp2p PubSub](https://github.com/libp2p/specs/tree/master/pubsub)
+- Translating an IPNS record name to/from a PubSub topic
+- Layering persistence onto libp2p PubSub
+
+# Translating an IPNS record name to/from a PubSub topic
+
+For a given IPNS local record key described in the IPNS Specification the PubSub topic is:
+
+**Topic format:** `/record/base64url-unpadded(key)`
+
+where base64url-unpadded is an unpadded base64url as specified in [IETF RFC 4648](https://tools.ietf.org/html/rfc4648)
+
+# Layering persistence onto libp2p PubSub
+
+libp2p PubSub does not have any notion of persistent data built into it. However, we can layer persistence on top of PubSub by utilizing [libp2p Fetch](https://github.com/libp2p/specs/).
+
+The protocol has the following steps:
+1. Start State: Node `A` subscribes to the PubSub topic `t` corresponding to the local IPNS record key `k`
+2. `A` notices that a node `B` has connected to it and subscribed to `t`
+3. Some time passes (might be 0 seconds, or could use a more complex system to determine the duration)
+4. `A` sends `B` a Fetch request for `k`
+5. If Fetch returns a record that supersedes `A`'s current record then `A` updates its record and Publishes it to the network
+
+Note: PubSub does not guarantee that the a message sent by a peer `A` will be received by a peer `B` and it's possible
+(e.g. in systems like [gossipsub](https://github.com/libp2p/specs/tree/master/pubsub/gossipsub))
+that this is true even if `A` and `B` are already connected. Therefore, whenever `A` notices **any** node that has
+connected to it and subscribed to `t` it should run the Fetch protocol as described above. However, developers may have routers
+with properties that allow the amount of time in step 3 to increase arbitrarily large (including infinite) amounts.
+
+# Protocol
+
+A node `A` putting and getting updates to an IPNS key `k`, with computed PubSub topic `t`
+
+1. PubSub subscribe to `t`
+2. Run the persistence protocol, both to fetch data and return data to those that request it
+3. When updating a record do a PubSub Publish and keep the record locally
+4. When receiving a record if it's better than the current record keep it and republish the message
+5. (Optional) Periodically republish the best record available
+
+Note: 5 is optional because it is not necessary. However, receiving duplicate records are already handled efficiently

```

Comment:
```suggestion
Note: 5 is optional because it is not necessary for the basic functionality. It can be an opt-in for IPNS names user wants to keep alive (e.g. by pinning an IPNS path). 

Note: Receiving duplicate records are already handled efficiently
```

Reply:
> Worth mentioning how being optional makes it a very useful feature:

I don't think it adds any utility if the `Fetch` protocol is being called properly since any time a node connects to another one those two nodes will use `Fetch` to synchronize their state and propagate any updates they receive. So when a node gets a new peer it should synchronize with that peer's state. Because their should never be a time (aside from at the very beginning) when two neighboring nodes are out of sync then the entire connected network should be in sync.

Reply:
Right. I guess this spec is not a good place for mentioning "IPNS pinning/following".  The spec for multiple resolvers (https://github.com/ipfs/specs/issues/198) will be a better place for it.

Reply:
True, but also the rebroadcast functionality isn't actually necessary for pinning/following since the use of `Fetch` + propagating any fetched updates is essentially a more efficient and faster version of rebroadcast.